src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCaptureCollection.cs

```diff
@@ -47,6 +50,25 @@ public int Count
         }
 
         /// <summary>
+        /// Copies all the elements of the collection to the given array
+        /// beginning at the given index.
+        /// </summary>
+        public void CopyTo(Capture[] array, int arrayIndex)
+        {
+            if (array == null)
+                throw new ArgumentNullException("array");
+            if (arrayIndex < 0 || arrayIndex > array.Length)
```

Comment:
There's actually even a behavioral difference between `List<T>.CopyTo` and `Dictionary<TKey, TValue>.CopyTo`.

The following tests pass:

```c#
ICollection<int> list = new List<int> { 0, 1 };
int[] listArray = new int[2];
Assert.Throws<ArgumentOutOfRangeException>(() => list.CopyTo(listArray, -1));
Assert.Throws<ArgumentException>(() => list.CopyTo(listArray, 2));
Assert.Throws<ArgumentException>(() => list.CopyTo(listArray, 3));

ICollection<KeyValuePair<int, int>> dictionary = new Dictionary<int, int> { { 0, 0 }, { 1, 1 } };
KeyValuePair<int, int>[] dictionaryArray = new KeyValuePair<int, int>[2];
Assert.Throws<ArgumentOutOfRangeException>(() => dictionary.CopyTo(dictionaryArray, -1));
Assert.Throws<ArgumentException>(() => dictionary.CopyTo(dictionaryArray, 2));
Assert.Throws<ArgumentOutOfRangeException>(() => dictionary.CopyTo(dictionaryArray, 3));
```

Notice the last case is `ArgumentException` for `List<T>` but `ArgumentOutOfRangeException` for `Dictionary<TKey, TValue>`

Right now `CaptureCollection` and `GroupCollection` are consistent with `Dictionary<TKey, TValue>` whereas `MatchCollection` is consistent with `List<T>` (because `MatchCollection` uses `List<T>.CopyTo` internally).

Maybe it'd be better if I remove the `arrayIndex > array.Length` check altogether from `CaptureCollection` and `GroupCollection`, so the exceptions are consistent with `MatchCollection` and `List<T>`. Or not worry about it too much because it's all `ArgumentException` anyway?